README
DESCRIPTION
    Base class for handling the import of the spreadsheet file

    Subclass this.

SYNOPSIS
    # Your Importer class package My::Importer;

    use Moose;

    extends 'Data::Importer';

    sub mandatory_columns { return [qw/name description price/]; }

    sub validate_row { my ($self, $row, $lineno) = @_; # XXX validation #
    return $self->add_error("Some error at $lineno") if some condition; #
    return $self->add_warning("Some warning at $lineno") if some condition;
    $self->add_row($row); }

    sub import_row { my ($self, $row) = @_;
    $schema->resultset('Table')->create($row) or die; }

    # Import the file like this

    my $import = My::Importer->new( schema => $schema, file_name =>
    $file_name, ); $import->do_work;

ATTTRIBUTES
  schema
    Yes, we use DBIx::Class. Send your schema.

  dry_run
    Set to true if this is only a test

  file_name
    The name of the spreadsheet

  import_type
    Three formats are supported, csv, xls, ods

  mandatory
    Required input columns

  optional
    Required input columns

"PRIVATE" ATTRIBUTES
  import_iterator
    An iterator for reading the import file

  rows
    An arrayref w/ the rows to be inserted / updated

  warnings
    An arrayref w/ all the warnings picked up during processing

    These methods are associated with warnings:

   all_warnings
    Returns all elements

   add_warning
    Add a warning

   join_warnings
    Join all warnings

   count_warnings
    Returns the number of warnings

   has_warnings
    Returns true if there are warnings

   has_no_warnings
    Returns true if there isn't any warning

   clear_warnings
    Clear out all warnings

  errors
    An arrayref w/ all the errors picked up during processing

    These methods are associated with errors:

   all_errors
    Returns all elements

   add_error
    Add a error

   join_errors
    Join all errors

   count_errors
    Returns the number of errors

   has_errors
    Returns true if there are errors

   has_no_errors
    Returns true if there isn't any error

   clear_errors
    Clear out all errors

  timestamp
    Time of the import run.

METHODS
  mandatory_columns
    Builds the mandatory attribute, which gives an arrayref with the
    required columns.

  optional_columns
    Builds the optional attribute, which gives an arrayref with the optional
    columns.

  do_work
    Import a file

  validate
    Validates the import file

  validate_row
    Handle each individual row.

    This has to be written in the subclass

  import_run
    Imports the nodes after validate has been run

  import_transaction
    Called inside the transaction block.

    This is the method to add method modifiers to if processing before or
    after the row import loop is necessary.

  import_row
    Handle each individual row.

    This has to be written in the subclass

  import_failed
    Called if the import failed for some reason

  import_succeeded
    Called after the import has finished succesfully

  report
    Make a report of what happened



Changes
Revision history for Data::Importer

0.005     2014-09-02 19:44:28+02:00 Europe/Copenhagen
 - Add a Makefile.PL
 - Option to encode values
 - Use Spreadsheet::ReadSXC for ods

0.004     2014-07-30 16:49:58+02:00 Europe/Copenhagen
 - Pod
 - Specify overridable methods
 - Add mandatory and optional attributes. Mandatory fields are used to check with the actual file to see if
   it contains the necessary columns.
   For now the optional attribute is, well, optional.
 - Make xls iterator more failsafe

0.003     2014-07-10 22:38:00+02:00 Europe/Copenhagen
 - Rework the import part to also be row based, like the validate step.

0.002     2014-07-10 18:38:49CEST+0200 Europe/Copenhagen
 - Add missing dependencies to the makefile
 - Apparantly csv->fields is not consistent over Perl releases. Fixed

0.001     2014-07-09 19:32:50CEST+0200 Europe/Copenhagen
 - Initial release


t/lib/Test/Schema.pm
#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Test::Schema;
use base qw/DBIx::Class::Schema/;
__PACKAGE__->load_namespaces;

1;


t/lib/Test/Import/Products.pm
#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Test::Import::Products;

use 5.010;
use namespace::autoclean;
use Moose;
use MooseX::Traits;

extends 'Data::Importer';

=head1 METHODS

=head2 handle_row

Validates input

Called for each row in the input file

=cut

sub validate_row {
	my ($self, $row, $lineno) = @_;
	#product
	my %prow = ( %$row );
	$self->add_row(\%prow);
}

=head2 import_row

Performs the actual database update

=cut

sub import_row {
	my ($self, $row) = @_;
	next unless $row->{item_name};

	my $schema = $self->schema;

	my $data = {
		name => $row->{item_name},
		ingredients => $row->{ingredients},
		qty => $row->{qty},
	};
	$schema->resultset('Product')->create($data) or die;
}

__PACKAGE__->meta->make_immutable;

#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#

__END__


t/lib/Test/Schema/Result/Product.pm
#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Test::Schema::Result::Product;
use base qw/DBIx::Class::Core/;
__PACKAGE__->table('product');
__PACKAGE__->add_columns(qw/ id name ingredients qty /);
__PACKAGE__->set_primary_key('id');

1;


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.019.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Framework to import row-based (spreadsheet-, csv-) files into a database",
  "dist_author" => [
    "Kaare Rasmussen <kaare\@cpan.org>"
  ],
  "dist_name" => "Data-Importer",
  "dist_version" => "0.005",
  "license" => "perl",
  "module_name" => "Data::Importer",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "DateTime" => 0,
    "DateTime::Format::Pg" => 0,
    "Encode" => 0,
    "Moose" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::Traits" => 0,
    "Spreadsheet::ParseExcel" => 0,
    "Spreadsheet::ReadSXC" => 0,
    "Text::CSV" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.010"
  },
  "script_files" => [],
  "test_requires" => {
    "DBD::SQLite" => 0,
    "DBIx::Class::Core" => 0,
    "DBIx::Class::Schema" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0,
    "base" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  }
);


my %fallback_build_requires = (
  "DBD::SQLite" => 0,
  "DBIx::Class::Core" => 0,
  "DBIx::Class::Schema" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.3601",
  "Test::More" => 0,
  "base" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Framework to import row-based (spreadsheet-, csv-) files into a database",
  "AUTHOR" => "Kaare Rasmussen <kaare\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Data-Importer",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Data::Importer",
  "PREREQ_PM" => {
    "DateTime" => 0,
    "DateTime::Format::Pg" => 0,
    "Encode" => 0,
    "Moose" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::Traits" => 0,
    "Spreadsheet::ParseExcel" => 0,
    "Spreadsheet::ReadSXC" => 0,
    "Text::CSV" => 0,
    "namespace::autoclean" => 0
  },
  "TEST_REQUIRES" => {
    "DBD::SQLite" => 0,
    "DBIx::Class::Core" => 0,
    "DBIx::Class::Schema" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0,
    "base" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.005",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBD::SQLite" => 0,
  "DBIx::Class::Core" => 0,
  "DBIx::Class::Schema" => 0,
  "DateTime" => 0,
  "DateTime::Format::Pg" => 0,
  "Encode" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.3601",
  "Moose" => 0,
  "Moose::Util::TypeConstraints" => 0,
  "MooseX::Traits" => 0,
  "Spreadsheet::ParseExcel" => 0,
  "Spreadsheet::ReadSXC" => 0,
  "Test::More" => 0,
  "Text::CSV" => 0,
  "base" => 0,
  "lib" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Data/Importer/Iterator.pm
#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Data::Importer::Iterator;
$Data::Importer::Iterator::VERSION = '0.005';
use 5.010;
use namespace::autoclean;
use Moose;

=head1 Description

Base class for handling the import of a file

Currently, there are iterators for L<CSV|Data::Importer::Iterator::Csv>, L<ODS|Data::Importer::Iterator::Ods>, and L<XLS|Data::Importer::Iterator::Xls>, 

=head1 ATTRIBUTES

=head2 lineno

The line counter

=cut

has 'lineno' => (
	is => 'ro',
	isa => 'Num',
	traits => ['Counter'],
	default => 0,
	handles => {
		inc_lineno => 'inc',
		reset_lineno => 'reset',
	},
);

=head2 file_name

The name of the import file

=cut

has file_name => (
	is => 'ro',
	isa => 'Str',
);

=head2 mandatory

Required input columns

=cut

has 'mandatory' => (
	is => 'ro',
	isa => 'ArrayRef',
);

=head2 optional

Required input columns

=cut

has 'optional' => (
	is => 'ro',
	isa => 'ArrayRef',
);

=head2 encoding

The encoding of the spreadsheet

=cut

has encoding => (
	is => 'ro',
	isa => 'Str',
	predicate => 'has_encoding',
);

=head1 "PRIVATE" ATTRIBUTES

=head2 file

The import file

=cut

has 'file' => (
	is => 'ro',
	lazy_build => 1,
);

=head1 METHODS

=head2 _build_file

The lazy builder for the file

The base class opens a file as UTF-8 and returns it.

=cut

sub _build_file {
	my $self = shift;
	my $filename = $self->file_name;
	open(my $file, "<:encoding(utf8)", $filename) or die "$filename: $!";

	return $file;
}

=head2 next

Return the next row of data from the file

=cut

sub next {
	return 'should be sub classed';
}

__PACKAGE__->meta->make_immutable;

#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#

__END__


lib/Data/Importer/Iterator/Csv.pm
#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Data::Importer::Iterator::Csv;
$Data::Importer::Iterator::Csv::VERSION = '0.005';
use 5.010;
use namespace::autoclean;
use Moose;
use Text::CSV;
use Encode qw/encode/;

extends 'Data::Importer::Iterator';

=head1 Description

Subclass for handling the import of a csv file

=head1 ATTRIBUTES

=head2 csv

The csv object

=cut

has 'csv' => (
	is => 'ro',
	lazy_build => 1,
);

=head1 "PRIVATE" ATTRIBUTES

=head2 file

The import file

=cut

has 'file' => (
	is => 'ro',
	lazy_build => 1,
);

=head1 METHODS

=head2 _build_csv

The lazy builder for the csv object

=cut

sub _build_csv {
	my $self = shift;
	my $csv = Text::CSV->new ({ binary => 1, eol => $/ });
	return $csv;
}

=head2 _build_file

The lazy builder for the file

The base class opens a file as UTF-8 and returns it.

=cut

sub _build_file {
	my $self = shift;
	my $filename = $self->file_name;
	open(my $file, "<:encoding(utf8)", $filename) or die "$filename: $!";

	return $file;
}

=head2 next

Return the next row of data from the file

=cut

sub next {
	my $self = shift;
	my $file = $self->file;
	my $csv = $self->csv;
	# Use the first row as column names:
	if (!$csv->column_names) {
		my $row = $csv->getline($file);
		my @fieldnames = map {my $header = lc $_; $header =~ tr/ /_/; $header} @$row;
		die "Only one column detected, please use comma ',' to separate data." if @fieldnames < 2;
		my %fieldnames = map {$_ => 1} @fieldnames;
		if (my @missing = grep {!$fieldnames{$_} } @{ $self->mandatory }) {
			die 'Column(s) required, but not found:' . join ', ', @missing;
		}

		$csv->column_names(@fieldnames);
	}
	$self->inc_lineno;
	my $line = $csv->getline_hr($file) or return;
	my $row = $self->has_encoding ? { map {$_ => encode($self->encoding, $line->{$_})} keys %$line } : $line;
	return $row;
}

__PACKAGE__->meta->make_immutable;

#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#

__END__


lib/Data/Importer/Iterator/Ods.pm
#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Data::Importer::Iterator::Ods;
$Data::Importer::Iterator::Ods::VERSION = '0.005';
use 5.010;
use namespace::autoclean;
use Moose;
use Encode qw(encode);
use Spreadsheet::ReadSXC qw/read_sxc/;

extends 'Data::Importer::Iterator';

=head1 Description

Subclass for handling the import of an ods file

=head1 ATTRIBUTES

=head2 ods

The ods object

=cut

has 'ods' => (
	is => 'ro',
	lazy_build => 1,
);

=head2 sheet

The sheet name or number

=cut

has sheet => (
	is => 'ro',
	isa => 'Int',
	default => 0,
);

=head2 options

Options to be passed to the Spreadsheet::ReadSXC constructor

=cut

has 'options' => (
	is => 'ro',
	isa => 'HashRef',
	default => sub {
		return {
			OrderBySheet => 1,
		};
	},
	lazy => 1,
);

=head1 "PRIVATE" ATTRIBUTES

=head2 column_names

The column names

=cut

has column_names => (
	is => 'rw',
	isa => 'ArrayRef',
	predicate => 'has_column_names',
);

=head1 METHODS

=head2 _build_ods

The lazy builder for the ods object

=cut

sub _build_ods {
	my $self = shift;
	my $worksheets_ref = read_sxc($self->file_name, $self->options);
	my $ods = $$worksheets_ref[0]{data};
	return $ods;
}

=head2 next

Return the next row of data from the file

=cut

sub next {
	my $self = shift;
	my $ods = $self->ods;
	# Use the first row as column names:
	if (!$self->has_column_names) {
		my @fieldnames = map {my $header = lc $_; $header =~ tr/ /_/; $header} @{ $ods->[$self->lineno] };
		die "Only one column detected, please use comma ',' to separate data." if @fieldnames < 2;
		my %fieldnames = map {$_ => 1} @fieldnames;
		if (my @missing = grep {!$fieldnames{$_} } @{ $self->mandatory }) {
			die 'Column(s) required, but not found:' . join ', ', @missing;
		}

		$self->column_names(\@fieldnames);
	}
	my $columns = $self->column_names;
	$self->inc_lineno;
	my @cells = @{ $ods->[$self->lineno] || [] };
	return unless grep { $_ } @cells;

	my $colno = 0;
    @cells = map { encode($self->encoding, $_) } @cells if $self->has_encoding;
    return { map { $columns->[$colno++] => $_ } @cells };
}

__PACKAGE__->meta->make_immutable;

#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#

__END__


lib/Data/Importer/Iterator/Xls.pm
#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Data::Importer::Iterator::Xls;
$Data::Importer::Iterator::Xls::VERSION = '0.005';
use 5.010;
use namespace::autoclean;
use Moose;
use Encode qw(encode);
use Spreadsheet::ParseExcel;

extends 'Data::Importer::Iterator';

=head1 Description

Subclass for handling the import of an excel file

=head1 ATTRIBUTES

=head2 excel

The excel object

=cut

has 'excel' => (
	is => 'ro',
	lazy_build => 1,
);

=head2 sheet

The sheet name or number

=cut

has sheet => (
	is => 'ro',
	isa => 'Str',
	default => 0,
);

=head1 "PRIVATE" ATTRIBUTES

=head2 column_names

The column names

=cut

has column_names => (
	is => 'rw',
	isa => 'ArrayRef',
	predicate => 'has_column_names',
);

=head1 METHODS

=head2 _build_excel

The lazy builder for the excel object

=cut

sub _build_excel {
	my $self = shift;
	my $table = Spreadsheet::ParseExcel->new()
		->parse($self->file_name)
		->worksheet($self->sheet);
	return $table;
}

=head2 next

Return the next row of data from the file

=cut

sub next {
	my $self = shift;
	my $xls = $self->excel;
	state $rc = [$xls->row_range];
	state $cc = [$xls->col_range];
	# Use the first row as column names:
	if (!$self->has_column_names) {
		my @fieldnames = map {my $header = lc $_; $header =~ tr/ /_/; $header} grep {$_} $self->_get_row_values($xls, @$cc);
		die "Only one column detected, please use comma ',' to separate data." if @fieldnames < 2;
		my %fieldnames = map {$_ => 1} @fieldnames;
		if (my @missing = grep {!$fieldnames{$_} } @{ $self->mandatory }) {
			die 'Column(s) required, but not found:' . join ', ', @missing;
		}

		$self->column_names(\@fieldnames);
		$cc->[1] = scalar @fieldnames;
	}
	$self->inc_lineno;
	return $self->_get_row($xls, @$cc);
}

sub _get_row_values {
	my ($self, $xls, $from, $to) = @_;
	my @cells;
	push @cells, $xls->get_cell($self->lineno, $_)->value for $from..$to;
	return @cells;
}

sub _get_row {
	my ($self, $xls, $from, $to) = @_;
	my $colnames = $self->column_names;
	my %cells;
	my $cells;
	for my $colno ($from..$to) {
		my $colname = $colnames->[$colno - $from] // '';
		if (my $cell = $xls->get_cell($self->lineno, $colno)) {
			my $value = $cell->value;
			if ($cell->encoding == 2) {
				$value = encode('utf-16BE', $cell->value);
			} elsif ($self->has_encoding) {
				$value = encode($self->encoding, $value);
			}
			$cells{ $colname } = $value;
			$cells++;
		} else {
			$cells{ $colname } = undef;
		}
	}
	return $cells ? \%cells : undef;
}

__PACKAGE__->meta->make_immutable;

#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#

__END__


lib/Data/Importer.pm
#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Data::Importer;
$Data::Importer::VERSION = '0.005';
use 5.010;
use namespace::autoclean;
use Moose;
use Moose::Util::TypeConstraints;
use DateTime;
use DateTime::Format::Pg;

with 'MooseX::Traits';

# ABSTRACT: Framework to import row-based (spreadsheet-, csv-) files into a database

=head1 DESCRIPTION

Base class for handling the import of the spreadsheet file

Subclass this.

=head1 SYNOPSIS

# Your Importer class
package My::Importer;

use Moose;

extends 'Data::Importer';

sub mandatory_columns {
	return [qw/name description price/];
}

sub validate_row {
	my ($self, $row, $lineno) = @_;
	# XXX validation
	# return $self->add_error("Some error at $lineno") if some condition;
	# return $self->add_warning("Some warning at $lineno") if some condition;
	$self->add_row($row);
}

sub import_row {
	my ($self, $row) = @_;
	$schema->resultset('Table')->create($row) or die;
}

# Import the file like this

my $import = My::Importer->new(
	schema => $schema,
	file_name => $file_name,
);
$import->do_work;

=head1 ATTTRIBUTES

=head2 schema

Yes, we use DBIx::Class. Send your schema.

=cut

has 'schema' => (
	isa => 'DBIx::Class',
	is => 'ro'
);

=head2 dry_run

Set to true if this is only a test

=cut

has 'dry_run' => (
	isa => 'Bool',
	is => 'ro',
	default => 0
);

=head2 file_name

The name of the spreadsheet

=cut

has file_name => (
	is => 'ro',
	isa => 'Str',
);

=head2 encoding

The encoding of the spreadsheet

=cut

has encoding => (
	is => 'ro',
	isa => 'Str',
	default => 'utf8',
	lazy => 1,
);

=head2 import_type

Three formats are supported, csv, xls, ods

=cut

enum 'import_types', [qw(csv ods xls)];

has 'import_type' => (
	is => 'ro',
	isa => 'import_types',
	lazy_build => 1,
);

=head2 mandatory

Required input columns

=cut

has 'mandatory' => (
	is => 'ro',
	isa => 'ArrayRef',
	lazy_build => 1,
	builder => 'mandatory_columns'
);

=head2 optional

Required input columns

=cut

has 'optional' => (
	is => 'ro',
	isa => 'ArrayRef',
	lazy_build => 1,
	builder => 'optional_columns'
);

=head1 "PRIVATE" ATTRIBUTES

=head2 import_iterator

An iterator for reading the import file

=cut

has 'import_iterator' => (
	is => 'ro',
	lazy_build => 1,
);

=head2 rows

An arrayref w/ the rows to be inserted / updated

=cut

has 'rows' => (
	is => 'ro',
	isa => 'ArrayRef',
	traits => [qw/Array/],
	handles => {
		add_row => 'push',
	},
	default => sub { [] },
);

=head2 warnings

An arrayref w/ all the warnings picked up during processing

These methods are associated with warnings:

=head3 all_warnings

Returns all elements

=head3 add_warning

Add a warning

=head3 join_warnings

Join all warnings

=head3 count_warnings

Returns the number of warnings

=head3 has_warnings

Returns true if there are warnings

=head3 has_no_warnings

Returns true if there isn't any warning

=head3 clear_warnings

Clear out all warnings

=cut

has 'warnings' => (
	traits  => ['Array'],
	is	  => 'ro',
	isa	 => 'ArrayRef[Str]',
	default => sub { [] },
	handles => {
		all_warnings	=> 'elements',
		add_warning	 => 'push',
		join_warnings   => 'join',
		count_warnings  => 'count',
		has_warnings	=> 'count',
		has_no_warnings => 'is_empty',
		clear_warnings  => 'clear',
	},
);

=head2 errors

An arrayref w/ all the errors picked up during processing

These methods are associated with errors:

=head3 all_errors

Returns all elements

=head3 add_error

Add a error

=head3 join_errors

Join all errors

=head3 count_errors

Returns the number of errors

=head3 has_errors

Returns true if there are errors

=head3 has_no_errors

Returns true if there isn't any error

=head3 clear_errors

Clear out all errors

=cut

has 'errors' => (
	traits  => ['Array'],
	is	  => 'ro',
	isa	 => 'ArrayRef[Str]',
	default => sub { [] },
	handles => {
		all_errors	=> 'elements',
		add_error	 => 'push',
		join_errors   => 'join',
		count_errors  => 'count',
		has_errors	=> 'count',
		has_no_errors => 'is_empty',
		clear_errors  => 'clear',
	},
);

=head2 timestamp

Time of the import run.

=cut

has 'timestamp' => (
	isa => 'Str',
	is => 'rw',
	lazy => 1,
	default => sub {
		return DateTime::Format::Pg->format_datetime(DateTime->now);
	},
);

=head1 METHODS

=cut

sub _build_import_type {
	my $self = shift;
	my $file_name = $self->file_name;
	$file_name =~ /\.(\w+)$/ || return;

	return lc $1;
}

sub _build_import_iterator {
	my $self = shift;
	my $import_type = ucfirst $self->import_type;
	my $classname = "Data::Importer::Iterator::$import_type";
	eval "require $classname" or die $@;
	return $classname->new(
		file_name => $self->file_name,
		mandatory => $self->mandatory,
		optional  => $self->optional,
		encoding  => $self->encoding,
	);
}

=head2 mandatory_columns

Builds the mandatory attribute, which gives an arrayref with the required columns.

=cut

sub mandatory_columns { [] }

=head2 optional_columns

Builds the optional attribute, which gives an arrayref with the optional columns.

=cut

sub optional_columns { [] }

=head2 do_work

Import a file

=cut

sub do_work {
	my $self = shift;
	$self->validate;
	$self->import_run unless $self->dry_run or $self->has_errors;
	$self->report;
	return 1;
}

=head2 validate

Validates the import file

=cut

sub validate {
	my ($self) = @_;
	my $iterator = $self->import_iterator;

	while (my $row = $iterator->next) {
		my $row_res = $self->validate_row($row, $iterator->lineno);
	}
}

=head2 validate_row

Handle each individual row.

This has to be written in the subclass

=cut

sub validate_row {
	my ($self, $row, $lineno) = @_;
	die "You have to provide your own validate_row";
}

=head2 import_run

Imports the nodes after validate has been run

=cut

sub import_run {
	my ($self) = @_;
	die "Trying to run import even with errors!" if $self->has_errors;

	my $schema = $self->schema;
	my $transaction = sub {
		$self->import_transaction;
		return;
	};
	my $rs;
	eval {
		$rs = $schema->txn_do($transaction);
	};
	if ($@) {
		$self->import_failed($@);
	} else {
		$self->import_succeeded;
	}
}

=head2 import_transaction

Called inside the transaction block.

This is the method to add method modifiers to if processing before or after the
row import loop is necessary.

=cut

sub import_transaction {
	my $self = shift;
	my $schema = $self->schema;
	for my $row (@{ $self->rows }) {
		$self->import_row($row);
	}
}

=head2 import_row

Handle each individual row.

This has to be written in the subclass

=cut

sub import_row {
	my ($self, $row, $lineno) = @_;
	die "You have to provide your own import_row";
}

=head2 import_failed

Called if the import failed for some reason

=cut

sub import_failed {
	my ($self, $error) = @_;
	die "Database error: $error \n";
}

=head2 import_succeeded

Called after the import has finished succesfully

=cut

sub import_succeeded { }

=head2 report

Make a report of what happened

=cut

sub report {
	my ($self) = @_;
	warn 'Warnings: ' . join("\n", @{ $self->warnings }) if $self->has_warnings;
	warn 'Errors ' . join("\n", @{ $self->errors }) if $self->has_errors;
}

__PACKAGE__->meta->make_immutable;

#
# This file is part of Data-Importer
#
# This software is copyright (c) 2014 by Kaare Rasmussen.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#

__END__
